src/win/util.c

```diff
@@ -1166,56 +1166,74 @@ int uv_os_homedir(char* buffer, size_t* size) {
   if (buffer == NULL || size == NULL || *size == 0)
     return UV_EINVAL;
 
-  /* Check if the USERPROFILE environment variable is set first */
-  path = malloc(*size * sizeof(WCHAR));
-
-  if (path == NULL)
-    return UV_ENOMEM;
+  bufsize = MAX_PATH;
 
-  len = GetEnvironmentVariableW(L"USERPROFILE", path, *size);
+  /* Check if the USERPROFILE environment variable is set first */
+  len = GetEnvironmentVariableW(L"USERPROFILE", path, bufsize);
 
   if (len == 0) {
     r = GetLastError();
-    free(path);
 
+    /* Don't return an error if USERPROFILE was not found */
     if (r != ERROR_ENVVAR_NOT_FOUND)
       return uv_translate_sys_error(r);
+  } else if (len > MAX_PATH) {
+    /* This should not be possible */
+    return UV_EIO;
   } else {
-    if (len > *size) {
-      free(path);
-      *size = len - 1;
+    /* Check how much space we need */
+    bufsize = uv_utf16_to_utf8(path, -1, NULL, 0);
+
+    if (bufsize == 0) {
+      return uv_translate_sys_error(GetLastError());
+    } else if (bufsize > *size) {
+      *size = bufsize - 1;
       return UV_ENOBUFS;
     }
 
+    /* Convert to UTF-8 */
     bufsize = uv_utf16_to_utf8(path, -1, buffer, *size);
-    assert(len + 1 == bufsize);
-    free(path);
-    *size = len;
 
+    if (bufsize == 0)
+      return uv_translate_sys_error(GetLastError());
+
+    *size = bufsize - 1;
     return 0;
   }
 
-  /* USERPROFILE is not set, so call SHGetKnownFolderPath() */
-  if (SHGetKnownFolderPath(&FOLDERID_Profile, 0, NULL, &path) != S_OK)
+  /* USERPROFILE is not set, so call GetUserProfileDirectoryW() */
+  if (OpenProcessToken(GetCurrentProcess(), TOKEN_READ, &token) == 0)
     return uv_translate_sys_error(GetLastError());
 
-  bufsize = uv_utf16_to_utf8(path, -1, buffer, *size);
-
-  if (bufsize == 0) {
+  if (!GetUserProfileDirectoryW(token, path, &bufsize)) {
```

Comment:
maybe assing bufsize here? I guess we don't really need to do it earlier, since it's not an in-out parameter.